src/toil/lib/threading.py

```diff
@@ -335,9 +333,222 @@ def process_name_exists(workDir, name):
+        
+        # We need a mutex name to allow only one process to be entering or
+        # leaving at a time.
+        self.mutex = name + '-arena-lock'
+        
+        # We need a way to track who is actually in, and who was in but died.
+        # So everybody gets a locked file (again).
+        # TODO: deduplicate with the similar logic for process names, and also
+        # deferred functions.
+        self.lockfileDir = os.path.join(workDir, name + '-arena-members')
+        
+        # When we enter the arena, we fill this in with the FD of the locked
+        # file that represents our presence.
+        self.lockfileFD = None
+        # And we fill this in with the file name
+        self.lockfileName = None
+        
+    def enter(self):
+        """
+        This process is entering the arena. If cleanup is in progress, blocks
+        until it is finished.
+        
+        You may not enter the arena again before leaving it.
+        """
+       
+        log.debug('Joining arena %s', self.lockfileDir)
+       
+        # Make sure we're not in it already.
+        assert self.lockfileName is None
+        assert self.lockfileFD is None
+        
+        with global_mutex(self.workDir, self.mutex):
+            # Now nobody else should also be trying to join or leave.
+            
+            try:
+                # Make sure the lockfile directory exists.
+                os.mkdir(self.lockfileDir)
+            except FileExistsError:
+                pass
+            
+            # Make ourselves a file in it and lock it to prove we are alive.
+            self.lockfileFD, self.lockfileName = tempfile.mkstemp(dir=self.lockfileDir)
+            # Nobody can see it yet, so lock it right away
+            fcntl.lockf(self.lockfileFD, fcntl.LOCK_EX)
+            
+            # Now we're properly in, so release the global mutex
+            
+        log.debug('Now in arena %s', self.lockfileDir)
+        
+    def leave(self):
+        """
+        This process is leaving the arena. If this process happens to be the
+        last process standing, yields something, with other processes blocked
+        from joining the arena until the loop body completes and the process
+        has finished leaving. Otherwise, does not yield anything.
+        
+        Should be used in a loop:
+        
+            for _ in arena.leave():
+                # If we get here, we were the last process. Do the cleanup
+                pass
+        """
+        
+        # Make sure we're in it to start.
+        assert self.lockfileName is not None
+        assert self.lockfileFD is not None
+        
+        log.debug('Leaving arena %s', self.lockfileDir)
+        
+        with global_mutex(self.workDir, self.mutex):
+            # Now nobody else should also be trying to join or leave.
+            
+            # Take ourselves out.
+            try:
+                os.unlink(self.lockfileName)
+            except:
```

Comment:
What happens if this has an error and cannot delete the lock file?

Is this worth posting a debug log message when this occurs?

Reply:
I figured unlink would only error out if the file was already gone. But it is possible the delete will somehow not be allowed.

In that case I think the file just gets left behind; the impact is that everyone else who leaves will see that the file is there, see that it isn't locked, try to delete it, fail to delete it, and continue on. There will be a lot of wasted time attempting deletes, and all the files will be left behind when the workflow finishes, but I don't think anything will actually *break*.

We could print a warning, but it would be a warning along the lines of "the filesystem doesn't work". I wouldn't add a debug message here unless we actually go to debug a problem where the workdir isn't properly being cleaned.